app/models/reduction_fetcher.rb

```diff
@@ -35,8 +38,12 @@ def reduce_by_subject?
     @topic == :reduce_by_subject
   end
 
-  def where(query)
-    return @subject_reductions.where(query.except(:user_id)) if reduce_by_subject?
-    return @user_reductions.where(query.except(:subject_id)) if reduce_by_user?
+  def locate_in_place(selector, relation, factory)
+    match = relation.to_a.map(&:serializable_hash).select{ |record| key_match(record, selector) }
+    match.empty? ? factory.new(selector) : match[0]
```

Comment:
Minor - If you are just looking for the first match if it exists, you could swap the map for a `find` https://ruby-doc.org/core-2.6.3/Enumerable.html#method-i-detect

Worst case is still O(N) but the best case is less than that.